Dev/event handler

jo-bru opened the pull request:
This version of the controller firmware is not complete yet, as discussed.
This version includes the following changes/implementations:

**To be able to handle different sorts of feedback messages, a `ResponseCode` field with the following options was included:**

- `DEBUG:` used only for debugging purpose 

- `ERROR`: used for error handling

- `ACK`: used for acknowledging actions that are not finished yet (used for non-blocking/event-driven mode)

- `DONE`: used for giving feedback on finished actions 

- `DATA`: used for returning data to the gateway

**For handling actions and events on the gateway, the following `ProfileState`s are used to indicate the state of each profile:**

- `UNREG`: the profile is created, but not yet registered on the MCU

- `IDLE`: the profile is registered and available for further requests

- `WAIT`: the profile is waiting for an ACK (blocking state: if a profile is in the WAIT state all profiles of the same MCU are blocked)

- `BUSY`: the profile is waiting for a DONE/DATA (non-blocking state: other profiles are not blocked)

**To manage all profile registered on the MCU, the `profile_manager.cpp` sub-module was added (#7):**

- in the `registered_profiles[255]` are all registered profiles saved to retrieve information for each profile 

- an existing profile gets simply overwritten

- for now, the deletion of profiles is not supported

- every successful registration is acknowledged with a DONE feedback message

**The following drivers were added/completed (#9):**

- `digital_generic`:  functions for reading a digital pin in blocking and non-blocking (event-driven) mode were added

- `color_sensor`: for now, the IIC address of the device is hardcoded, therefore only one color sensor can be used at the same time

-  `ultrasonic_sensor`: for now, the measurement is hardcoded (only for testing, as I have no ultrasonic sensor at home)

- `step_lowlevel`: not implemented yet (only skeleton created with the automatic driver initialization)

**To test the firmware, the following tests were added to the `simple_gateway.py` file:**

- Updating profiles: overwrite profile 2 every 5 loops the change LED color (green <=> blue)

- Event handling: create a profile for button A and listen for an event (`read_event_digital()`)

- Color sensor: get the current measurement of the sensor (only working if the sensor is connected to IIC port => **!! wrong label on controller backside: use Digital ports for IIC**)

- Ultrasonic sensor: get a measurement of the sensor (result is hardcoded as explained before)

**The `event_handler()` was added to handle non-blocking requests (#5 ):**
- each driver that should support non-blocking requests needs an `event_driver_name` function (e.g. `event_digital_generic()`)

**The automatic driver initialization script is completed (#8):**

- when running the script as described in the README file, all needed code sections are automatically filled in marked with TODO where manual implementation is needed

jo-bru committed:
start with profile_manager: registered profile array => not working properly

jo-bru committed:
small changes in request_handler() for debugging

iomz committed:
Sync with the main

jo-bru committed:
Merge dev/protobuf-helper on dev/profile_manager

jo-bru committed:
Implemented better delete_profile function

jo-bru committed:
Create event_handler for all non-blocking actions

jo-bru committed:
Merge main to dev/event-handler (new: protobuf_helper)

jo-bru committed:
Add feedback codes and profile states

jo-bru committed:
Add test for overlapping registrations

jo-bru committed:
Add profile states in .py + Add digital event handling

jo-bru committed:
Extend automatic driver init + Add driver skeletons: (color_sensor, ultrasonic_sensor, step_lowlevel)

jo-bru committed:
Add simple color sensor driver implementation (supports only 1 sensor)

jo-bru committed:
Add Ultrasonic sensor driver implementation (not tested yet!)

iomz commented on protobuf/line_protocol.proto:
```diff
@@ -24,17 +32,16 @@ enum UartPort {
 /*========================================================================*/
 
 // message sent from controller to gateway
 message Feedback {
```
Now you have `ResponseCode` for typing the message from MCU to Gateway. Why not renaming this message to `Response`? – the name will then correspond to the Request/Response interactions pattern which is the basis of Hypermedia Control achieved by and throughout our system.

jo-bru replied:
Done.

iomz commented on protobuf/line_protocol.proto:
```diff
@@ -24,17 +32,16 @@ enum UartPort {
-  uint32 profile_id = 1;  // used to identify profile
-  string message = 2;     // used for debugging messages
-  bytes data = 3;
+  ResponseCode code = 1; // used for feedback message handling
```
If we rename `Feedback` to `Response`, this field name fits right in (i.e., `Response.code`). Otherwise, `Feedback.code` is actually `ResponseCode` enum, and that's a disparity.

jo-bru replied:
Done with commit 50cdd1c.

iomz commented on protobuf/line_protocol.proto:
```diff
@@ -24,17 +32,16 @@ enum UartPort {
+  ResponseCode code = 1; // used for feedback message handling
+  uint32 profile_id = 2; // used to identify profile
+  string message = 3;    // used for debugging messages
+  bytes data = 4;        // holds all response data
```
Now we have five different types. My understanding is the following:

- `message` will be only  used for `DEBUG`
- `data` conveys necessary, functional information for the request made.

Why can't we have a single `bytes` field called `payload` that can be used for all the message types you defined?

With this, also `DONE` and `DATA` can be merged into one message type – `DONE` is a `DATA` message with an empty `payload`.

jo-bru replied:
Done with commit: 4f9e811.

iomz committed:
Purge the vendor files and added through PlatformIO

iomz commented on src/drivers/color_sensor.cpp:
```diff
@@ -0,0 +1,60 @@
+#include "color_sensor.h"
+#include "helper_files/Adafruit_TCS34725.h"
```
This is called a "vendor" source, any third-party codebase/library/package is supposed to be treated in a way that others do not misunderstand it is part of "your" code, mainly to respect the license – the TCS34725 is BSD-licensed.

See my commit 6cf4568 fixes this.

iomz reviewed (commented):
@jo-bru As requested, I'm just leaving comments and neither approving nor requesting changes.

- `ResponseCode` for `Feedback`: See my in-line comments.
- `ProfileManager`: Why don't you create a `ProfileManager` class and do the management in an object-oriented manner instead of `extern Registration[]`?
- `ProfileState` and `[store|delete]_profile()`: Does the FSM look like this? (the transition is slightly modified based on my recommendation) Isn't the `BUSY` and `WAIT` the other way around?
![Profile lifecycle?](https://user-images.githubusercontent.com/26181/97418916-b4951000-1909-11eb-92a5-bc64b4d26efa.jpg)
- ~~`Action` and `Registration` sub-messages with the [A|R]_types: Just with a quick glance today, I still couldn't see the reasoning for creating the two different sub-messages.~~
- `tools/driver_init`: It's great to leave a code snippet for templating the future component, but perhaps you would like existing templating engines/tools rather than sed+txt. (e.g., https://github.com/jinja2cpp/Jinja2Cpp)

jo-bru committed:
Change ProfileManager structure and Feedback => Response name

jo-bru committed:
Add working/tested uArm functionalities

jo-bru committed:
Add command list handling on gateway for uArm robots

jo-bru committed:
Add working/tested ultrasonic sensor handling

jo-bru committed:
Add tube sensor test + fix color sensor bug

jo-bru committed:
Add subroutine test (simple_gateway.py)

jo-bru committed:
Change profile FSM state names

jo-bru committed:
Simplify response message structure: message+data=>payload, remove DONE

jo-bru committed:
Add FSM + Caution section to README

jo-bru commented:
I've changed all the discussed parts and updated the FSM, which I added to the README file.
I've run a simple test (`subroutine_test()`) with the uArm robot at home (see GIF).

![subroutine_test](https://user-images.githubusercontent.com/67899546/98664289-79073680-234a-11eb-912a-a436a24fcfdd.gif)

Since the belt conveyor and the slider are too big to easily take home, I could not implement the step_lowlevel driver yet.
For now, I haven't implemented the non-volatile backup mechanism yet as it is labeled as an optional issue (see issue #3). 

The firmware should be ready to use with further gateway implementations IMO. We may have to optimize the firmware in the future to reduce the latency.

iomz commented:
@jo-bru Wonderful, thank you! I'll review the changes this afternoon. Let's discuss the gRPC gateway tomorrow.

iomz committed:
Remove the image for README from the repo

iomz commented on README.md:
```diff
@@ -28,6 +35,12 @@ To add a new Driver named <new_driver> run following command:
+- class skeleton in `simple_gateway.py`
+
+# Caution
+The labeling on the backside of the UFactory controller is not correct, IIC and digital ports are swapped. To connect an IIC device, use port 1 or port 2 indicated in the [uArm Controller user manual](https://cdn.shopify.com/s/files/1/0012/6979/2886/files/uArm_Controller_20190718.pdf?v=1597982243).
```
Nice :-)

iomz commented on src/profile_manager.h:
```diff
@@ -0,0 +1,27 @@
+
+#include "main.h"
+
+class ProfileManager
```
This is much cleaner.

iomz reviewed (approved):
I created some issues to be addressed outside this branch.

- #13
- #14
- #15 

Images for documentation should not go into the repository (6beed3290da778398bad0910367328c4e72399cf). Use GitHub's storage, for instance. (If you drag&drop image onto issue textbox, it starts uploading it and gives you a link)
